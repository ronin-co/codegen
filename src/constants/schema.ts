import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';

import type { TypeNode } from 'typescript';

import type { ModelField } from '@/src/types/model';

/**
 * A list of all model field types & their TypeScript type mapping.
 */
export const MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD = {
  blob: factory.createExpressionWithTypeArguments(
    factory.createPropertyAccessExpression(
      identifiers.ronin.namespace,
      identifiers.ronin.blob,
    ),
    undefined,
  ),
  boolean: factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
  date: factory.createTypeReferenceNode(identifiers.primitive.date),
  json: factory.createKeywordTypeNode(SyntaxKind.ObjectKeyword),
  link: factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
  number: factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
  string: factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
} satisfies Record<ModelField['type'], TypeNode>;

/**
 * An array of all possible query types as human readable strings.
 */
export const QUERY_TYPE_NAMES = [
  'ExtendedAdder',
  'ExtendedCounter',
  'ExtendedGetter',
  'ExtendedRemover',
  'ExtendedSetter',
] as const;

/**
 * A simple object mapping all query types to their human readable string.
 */
export const READABLE_QUERY_TYPE_NAMES = {
  ExtendedAdder: 'Add',
  ExtendedCounter: 'Count',
  ExtendedGetter: 'Get',
  ExtendedRemover: 'Remove',
  ExtendedSetter: 'Set',
} satisfies Record<(typeof QUERY_TYPE_NAMES)[number], string>;
