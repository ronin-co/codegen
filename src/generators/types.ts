import { SyntaxKind, addSyntheticLeadingComment, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';
import { MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD } from '@/src/constants/schema';
import { convertToPascalCase } from '@/src/utils/slug';

import type {
  InterfaceDeclaration,
  PropertySignature,
  TypeAliasDeclaration,
  TypeNode,
} from 'typescript';

import type { Model, ModelField } from '@/src/types/model';

const DEFAULT_FIELD_SLUGS = [
  'id',
  'ronin.createdAt',
  'ronin.createdBy',
  'ronin.locked',
  'ronin.updatedAt',
  'ronin.updatedBy',
] satisfies Array<string>;

/**
 * Generate all required type definitions for a provided RONIN model.
 *
 * This will generate an interface for the singular model type and a type alias
 * for the plural model type.
 *
 * The plural model type will be mapped to an array of the singular model type
 * and extend it with the plural model properties.
 *
 * @param models - All RONIN models of the addressed space.
 * @param model - A RONIN model to generate type definitions for.
 *
 * @returns - An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateTypes = (
  models: Array<Model>,
): Array<InterfaceDeclaration | TypeAliasDeclaration> => {
  const nodes = new Array<InterfaceDeclaration | TypeAliasDeclaration>();

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const modelIdentifier = factory.createIdentifier(
      convertToPascalCase(`${model.slug}Schema`),
    );
    const singularModelIdentifier = factory.createIdentifier(
      convertToPascalCase(model.slug),
    );
    const pluralSchemaIdentifier = factory.createIdentifier(
      convertToPascalCase(model.pluralSlug),
    );

    const mappedModelFields = fields
      .sort((a, b) => a.slug.localeCompare(b.slug))
      .map((field) => {
        const propertyUnionTypes = new Array<TypeNode>();

        switch (field.type) {
          case 'link': {
            const targetModel = models.find((model) => model.slug === field.target);
            propertyUnionTypes.push(
              targetModel
                ? factory.createTypeReferenceNode(
                    convertToPascalCase(`${targetModel.slug}Schema`),
                  )
                : factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
            );
            break;
          }
          case 'blob':
          case 'boolean':
          case 'date':
          case 'json':
          case 'number':
          case 'string': {
            propertyUnionTypes.push(MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD[field.type]);
            break;
          }
          default: {
            throw new Error('Unsupported field type found', {
              cause: field,
            });
          }
        }

        // If the field is not required, we need to mark it as `| null`.
        if (field.required === false)
          propertyUnionTypes.push(factory.createLiteralTypeNode(factory.createNull()));

        return factory.createPropertySignature(
          undefined,
          field.slug,
          undefined,
          factory.createUnionTypeNode(propertyUnionTypes),
        );
      })
      .filter(Boolean) as Array<PropertySignature>;

    /**
     * ```ts
     * interface SchemaSlugSchema extends ResultRecord {
     *    name: string | null;
     *    email: string;
     *    // ...
     * }
     * ```
     */
    const modelInterfaceDec = factory.createInterfaceDeclaration(
      undefined,
      modelIdentifier,
      [],
      [
        // All models should extend the `ResultRecord` interface.
        factory.createHeritageClause(SyntaxKind.ExtendsKeyword, [
          factory.createExpressionWithTypeArguments(
            identifiers.syntax.resultRecord,
            undefined,
          ),
        ]),
      ],
      mappedModelFields,
    );

    const modelSchemaName = factory.createTypeReferenceNode(modelIdentifier, []);

    /**
     * ```ts
     * export type SchemaSlug = SchemaSlugSchema;
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      singularModelIdentifier,
      undefined,
      modelSchemaName,
    );

    /**
     * ```ts
     * export type SchemaPluralSlug = Array<SchemaSlug>;
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      pluralSchemaIdentifier,
      undefined,
      factory.createTypeReferenceNode(identifiers.primitive.array, [modelSchemaName]),
    );

    // If the model does not have a summary / description
    // then we can continue to the next iteration & not add any comments.
    if (!model.summary) {
      nodes.push(modelInterfaceDec, singularModelTypeDec, pluralModelTypeDec);
      continue;
    }

    nodes.push(
      addSyntheticLeadingComment(
        modelInterfaceDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
      addSyntheticLeadingComment(
        singularModelTypeDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
      addSyntheticLeadingComment(
        pluralModelTypeDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
    );
  }

  return nodes;
};
