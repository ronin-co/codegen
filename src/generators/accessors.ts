import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';
import { convertToPascalCase } from '@/src/utils/slug';

import type { InterfaceDeclaration } from 'typescript';

import type { QUERY_TYPE_NAMES } from '@/src/constants/schema';

export const FUNCTION_PROPERTIES = [
  'apply',
  'arguments',
  'bind',
  'call',
  'caller',
  'length',
  'name',
  'prototype',
  'toString',
] satisfies Array<string>;

interface GenerateAccessorsOptions {
  slug: string;
  pluralSlug: string;
}

export interface GenerateAccessorsResult {
  singular: Record<(typeof QUERY_TYPE_NAMES)[number], InterfaceDeclaration>;
  plural: Record<(typeof QUERY_TYPE_NAMES)[number], InterfaceDeclaration>;
}

/**
 * Generates the accessors interface declarations for a given model.
 *
 * @param options - The options to generate accessors for.
 * @param options.slug - The singular slug of the model.
 * @param options.pluralSlug - The plural slug of the model.
 *
 * @returns The generated accessors interface declarations.
 */
export const generateAccessors = (
  options: GenerateAccessorsOptions,
): GenerateAccessorsResult => {
  const modelIdentifier = convertToPascalCase(`${options.slug}Schema`);
  const capitalizedSlug = convertToPascalCase(options.slug);
  const capitalizedPluralSlug = convertToPascalCase(options.pluralSlug);

  /**
   * ```ts
   * interface AccountAdder extends RONIN.IAdder<Account> {}
   * interface AccountsAdder extends RONIN.IAdder<Account> {}
   * ```
   */
  const createAdderInterface = (plural: boolean): InterfaceDeclaration => {
    const interfaceName = `${plural ? capitalizedPluralSlug : capitalizedSlug}Adder`;

    return factory.createInterfaceDeclaration(
      undefined,
      interfaceName,
      undefined,
      [
        factory.createHeritageClause(SyntaxKind.ExtendsKeyword, [
          factory.createExpressionWithTypeArguments(
            factory.createPropertyAccessExpression(
              identifiers.ronin.namespace,
              identifiers.ronin.accessor.adder,
            ),
            [factory.createTypeReferenceNode(modelIdentifier, [])],
          ),
        ]),
      ],
      [],
    );
  };

  /**
   * ```ts
   * interface AccountCounter extends RONIN.ICounter<Account> {}
   * interface AccountsCounter extends RONIN.ICounter<Account> {}
   * ```
   */
  const createCounterInterface = (plural: boolean): InterfaceDeclaration => {
    const interfaceName = `${plural ? capitalizedPluralSlug : capitalizedSlug}Counter`;

    return factory.createInterfaceDeclaration(
      undefined,
      interfaceName,
      undefined,
      [
        factory.createHeritageClause(SyntaxKind.ExtendsKeyword, [
          factory.createExpressionWithTypeArguments(
            factory.createPropertyAccessExpression(
              identifiers.ronin.namespace,
              identifiers.ronin.accessor.counter,
            ),
            [factory.createTypeReferenceNode(modelIdentifier, [])],
          ),
        ]),
      ],
      [],
    );
  };

  /**
   * ```ts
   * interface AccountGetter extends RONIN.IGetterSingular<Account> {}
   * interface AccountsGetter extends RONIN.IGetterPlural<Account> {}
   * ```
   */
  const createGetterInterface = (plural: boolean): InterfaceDeclaration => {
    const interfaceName = `${plural ? capitalizedPluralSlug : capitalizedSlug}Getter`;
    const accessorName = plural
      ? identifiers.ronin.accessor.getterPlural
      : identifiers.ronin.accessor.getterSingular;

    return factory.createInterfaceDeclaration(
      undefined,
      interfaceName,
      undefined,
      [
        factory.createHeritageClause(SyntaxKind.ExtendsKeyword, [
          factory.createExpressionWithTypeArguments(
            factory.createPropertyAccessExpression(
              identifiers.ronin.namespace,
              accessorName,
            ),
            [factory.createTypeReferenceNode(modelIdentifier, [])],
          ),
        ]),
      ],
      [],
    );
  };

  /**
   * ```ts
   * interface AccountRemover extends RONIN.IRemover<Account> {}
   * interface AccountsRemover extends RONIN.IRemover<Account> {}
   * ```
   */
  const createRemoverInterface = (plural: boolean): InterfaceDeclaration => {
    const interfaceName = `${plural ? capitalizedPluralSlug : capitalizedSlug}Remover`;

    return factory.createInterfaceDeclaration(
      undefined,
      interfaceName,
      undefined,
      [
        factory.createHeritageClause(SyntaxKind.ExtendsKeyword, [
          factory.createExpressionWithTypeArguments(
            factory.createPropertyAccessExpression(
              identifiers.ronin.namespace,
              identifiers.ronin.accessor.remover,
            ),
            [factory.createTypeReferenceNode(modelIdentifier, [])],
          ),
        ]),
      ],
      [],
    );
  };

  /**
   * ```ts
   * interface AccountSetter extends RONIN.ISetter<Account> {}
   * interface AccountsSetter extends RONIN.ISetter<Account> {}
   * ```
   */
  const createSetterInterface = (plural: boolean): InterfaceDeclaration => {
    const interfaceName = `${plural ? capitalizedPluralSlug : capitalizedSlug}Setter`;

    return factory.createInterfaceDeclaration(
      undefined,
      interfaceName,
      undefined,
      [
        factory.createHeritageClause(SyntaxKind.ExtendsKeyword, [
          factory.createExpressionWithTypeArguments(
            factory.createPropertyAccessExpression(
              identifiers.ronin.namespace,
              identifiers.ronin.accessor.setter,
            ),
            [factory.createTypeReferenceNode(modelIdentifier, [])],
          ),
        ]),
      ],
      [],
    );
  };

  return {
    singular: {
      ExtendedAdder: createAdderInterface(false),
      ExtendedCounter: createCounterInterface(false),
      ExtendedGetter: createGetterInterface(false),
      ExtendedRemover: createRemoverInterface(false),
      ExtendedSetter: createSetterInterface(false),
    },
    plural: {
      ExtendedAdder: createAdderInterface(true),
      ExtendedCounter: createCounterInterface(true),
      ExtendedGetter: createGetterInterface(true),
      ExtendedRemover: createRemoverInterface(true),
      ExtendedSetter: createSetterInterface(true),
    },
  };
};
