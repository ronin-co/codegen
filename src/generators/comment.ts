import { READABLE_QUERY_TYPE_NAMES } from '@/src/constants/schema';

import type { QUERY_TYPE_NAMES } from '@/src/constants/schema';

interface GenerateQueryTypeCommentResult {
  singular: string;
  plural: string;
}

/**
 * Generate a text comment for a provided model & a given query type.
 *
 * @param modelName - The name of the model to generate a comment for.
 * @param queryType - The query type to generate a comment for.
 *
 * @returns An object containing both the singular and plural comment strings.
 */
export const generateQueryTypeComment = (
  modelName: string,
  queryType: (typeof QUERY_TYPE_NAMES)[number],
): GenerateQueryTypeCommentResult => ({
  singular: ` ${READABLE_QUERY_TYPE_NAMES[queryType]} a single record of the ${modelName} model `,
  plural: ` ${READABLE_QUERY_TYPE_NAMES[queryType]} multiple records of the ${modelName} model `,
});
