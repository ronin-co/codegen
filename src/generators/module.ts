import { NodeFlags, SyntaxKind, addSyntheticLeadingComment, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';
import { QUERY_TYPE_NAMES } from '@/src/constants/schema';
import { generateQueryTypeComment } from '@/src/generators/comment';
import { convertToPascalCase } from '@/src/utils/slug';

import type {
  InterfaceDeclaration,
  ModuleDeclaration,
  Statement,
  TypeAliasDeclaration,
  TypeElement,
} from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Generate a module augmentation for the `ronin` module to override the
 * standard filter interfaces with ones that are correctly typed specific to
 * this space.
 *
 * @param models - An array of RONIN models to generate type definitions for.
 * @param schemas - An array of type declarations for the models.
 *
 * @returns A module augmentation declaration to be added to `index.d.ts`.
 */
export const generateModule = (
  models: Array<Model>,
  schemas: Array<InterfaceDeclaration | TypeAliasDeclaration>,
): ModuleDeclaration => {
  const moduleBodyStatements = new Array<Statement>();

  for (const schemaTypeDec of schemas) {
    moduleBodyStatements.push(schemaTypeDec);
  }

  /**
   * ```ts
   * declare const add: {};
   * declare const count: {};
   * declare const get: {};
   * declare const remove: {};
   * declare const set: {};
   * ```
   */
  for (const queryType of QUERY_TYPE_NAMES) {
    const declarationProperties = new Array<TypeElement>();
    for (const model of models) {
      const comment = generateQueryTypeComment(model.slug, queryType);
      const singularModelIdentifier = factory.createTypeReferenceNode(
        convertToPascalCase(model.slug),
      );
      const pluralSchemaIdentifier = factory.createTypeReferenceNode(
        convertToPascalCase(model.pluralSlug),
      );

      /**
       * ```ts
       * GetQuery[keyof GetQuery]
       * ```
       */
      const queryTypeValue = factory.createIndexedAccessTypeNode(
        factory.createTypeReferenceNode(
          // identifiers.compiler.queryType[queryType],
          identifiers.compiler.queryType[queryType],
          undefined,
        ),
        factory.createTypeOperatorNode(
          SyntaxKind.KeyOfKeyword,
          factory.createTypeReferenceNode(identifiers.compiler.queryType[queryType]),
        ),
      );

      /**
       * ```ts
       * account: DeepCallable<GetQuery[keyof GetQuery], Account | null>;
       * ```
       */
      const singularProperty = factory.createPropertySignature(
        undefined,
        model.slug,
        undefined,
        factory.createTypeReferenceNode(identifiers.syntax.deepCallable, [
          queryTypeValue,
          factory.createUnionTypeNode([
            singularModelIdentifier,
            factory.createLiteralTypeNode(factory.createNull()),
          ]),
        ]),
      );
      declarationProperties.push(
        addSyntheticLeadingComment(
          singularProperty,
          SyntaxKind.MultiLineCommentTrivia,
          comment.singular,
          true,
        ),
      );

      // TODO(@nurodev): Remove once RONIN officially supports
      // creating multiple records at once.
      if (queryType === 'add') continue;

      /**
       * ```ts
       * accounts: DeepCallable<GetQuery[keyof GetQuery], Array<Account>>;
       * ```
       */
      const pluralProperty = factory.createPropertySignature(
        undefined,
        model.pluralSlug,
        undefined,
        factory.createTypeReferenceNode(identifiers.syntax.deepCallable, [
          queryTypeValue,
          pluralSchemaIdentifier,
        ]),
      );
      declarationProperties.push(
        addSyntheticLeadingComment(
          pluralProperty,
          SyntaxKind.MultiLineCommentTrivia,
          comment.plural,
          true,
        ),
      );
    }

    const queryDeclaration = factory.createVariableStatement(
      [factory.createModifier(SyntaxKind.DeclareKeyword)],
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            queryType,
            undefined,
            factory.createTypeLiteralNode(declarationProperties),
          ),
        ],
        NodeFlags.Const,
      ),
    );

    moduleBodyStatements.push(queryDeclaration);
  }

  return factory.createModuleDeclaration(
    [factory.createModifier(SyntaxKind.DeclareKeyword)],
    identifiers.ronin.module.root,
    factory.createModuleBlock(moduleBodyStatements),
  );
};
