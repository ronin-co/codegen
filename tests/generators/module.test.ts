import { describe, expect, test } from 'bun:test';
import { model, string } from 'ronin/schema';

import { generateModule } from '@/src/generators/module';
import { printNodes } from '@/src/utils/print';

describe('module', () => {
  test('a basic model', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const moduleDeclaration = generateModule([AccountModel]);

    const moduleDeclarationStr = printNodes([moduleDeclaration]);

    expect(moduleDeclarationStr).toMatchSnapshot();
  });

  test('with no modules', () => {
    const moduleDeclaration = generateModule([]);

    const moduleDeclarationStr = printNodes([moduleDeclaration]);

    expect(moduleDeclarationStr).toMatchSnapshot();
  });

  test('with multiple models', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
    });

    const PostModel = model({
      slug: 'post',
      pluralSlug: 'posts',
      fields: {
        title: string({ required: true }),
        description: string(),
      },
    });

    // @ts-expect-error Codegen models types differ from the schema model types.
    const moduleDeclaration = generateModule([AccountModel, PostModel]);
    const moduleDeclarationStr = printNodes([moduleDeclaration]);
    expect(moduleDeclarationStr).toMatchSnapshot();
  });
});
