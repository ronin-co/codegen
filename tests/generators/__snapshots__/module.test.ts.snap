// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`module with no modules 1`] = `
"declare module "ronin" {
    declare const add: {};
    declare const count: {};
    declare const get: {};
    declare const remove: {};
    declare const set: {};
}
"
`;

exports[`module a basic model 1`] = `
"declare module "ronin" {
    declare const add: {
        /* Add a single account record */
        account: DeepCallable<AddQuery[keyof AddQuery], Account | null>;
    };
    declare const count: {
        /* Count a single account record */
        account: DeepCallable<CountQuery[keyof CountQuery], Account | null>;
        /* Count multiple account records */
        accounts: DeepCallable<CountQuery[keyof CountQuery], Accounts>;
    };
    declare const get: {
        /* Get a single account record */
        account: DeepCallable<GetQuery[keyof GetQuery], Account | null>;
        /* Get multiple account records */
        accounts: DeepCallable<GetQuery[keyof GetQuery], Accounts>;
    };
    declare const remove: {
        /* Remove a single account record */
        account: DeepCallable<RemoveQuery[keyof RemoveQuery], Account | null>;
        /* Remove multiple account records */
        accounts: DeepCallable<RemoveQuery[keyof RemoveQuery], Accounts>;
    };
    declare const set: {
        /* Set a single account record */
        account: DeepCallable<SetQuery[keyof SetQuery], Account | null>;
        /* Set multiple account records */
        accounts: DeepCallable<SetQuery[keyof SetQuery], Accounts>;
    };
}
"
`;

exports[`module with multiple models 1`] = `
"declare module "ronin" {
    declare const add: {
        /* Add a single account record */
        account: DeepCallable<AddQuery[keyof AddQuery], Account | null>;
        /* Add a single post record */
        post: DeepCallable<AddQuery[keyof AddQuery], Post | null>;
    };
    declare const count: {
        /* Count a single account record */
        account: DeepCallable<CountQuery[keyof CountQuery], Account | null>;
        /* Count multiple account records */
        accounts: DeepCallable<CountQuery[keyof CountQuery], Accounts>;
        /* Count a single post record */
        post: DeepCallable<CountQuery[keyof CountQuery], Post | null>;
        /* Count multiple post records */
        posts: DeepCallable<CountQuery[keyof CountQuery], Posts>;
    };
    declare const get: {
        /* Get a single account record */
        account: DeepCallable<GetQuery[keyof GetQuery], Account | null>;
        /* Get multiple account records */
        accounts: DeepCallable<GetQuery[keyof GetQuery], Accounts>;
        /* Get a single post record */
        post: DeepCallable<GetQuery[keyof GetQuery], Post | null>;
        /* Get multiple post records */
        posts: DeepCallable<GetQuery[keyof GetQuery], Posts>;
    };
    declare const remove: {
        /* Remove a single account record */
        account: DeepCallable<RemoveQuery[keyof RemoveQuery], Account | null>;
        /* Remove multiple account records */
        accounts: DeepCallable<RemoveQuery[keyof RemoveQuery], Accounts>;
        /* Remove a single post record */
        post: DeepCallable<RemoveQuery[keyof RemoveQuery], Post | null>;
        /* Remove multiple post records */
        posts: DeepCallable<RemoveQuery[keyof RemoveQuery], Posts>;
    };
    declare const set: {
        /* Set a single account record */
        account: DeepCallable<SetQuery[keyof SetQuery], Account | null>;
        /* Set multiple account records */
        accounts: DeepCallable<SetQuery[keyof SetQuery], Accounts>;
        /* Set a single post record */
        post: DeepCallable<SetQuery[keyof SetQuery], Post | null>;
        /* Set multiple post records */
        posts: DeepCallable<SetQuery[keyof SetQuery], Posts>;
    };
}
"
`;
